{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is an adaptation of the **RFdiffusion** Colab Implementation to run locally in Jupyter Notebook, preferably on the LRZ AI Cluster. It does not require any prior installations, you just have to keep in mind that all data is going to be saved in your home directory in a new directory called \"RFdiffusionJupyter\". So if you have RFdiffusion installed already you could for example save a little disk space by combining this directory with the original RFdiffusion directory. \n",
    "The original Colab Notebook can be found here:\n",
    "https://colab.research.google.com/github/sokrypton/ColabDesign/blob/main/rf/examples/diffusion_ori.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First thing to do is the installation and definition of functions. Don´t worry, if you already have installed it should skip installation. If you change any path names, etc. in your UI be sure to adapt in the code below! There might be problems with the PyTorch versions, so if all fails, just delete the directory RFdiffusionJupyter (after saving all your previous in and outputs as necessary!!!!) and run the installation cell again! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, subprocess, sys, shutil, time, random, string\n",
    "\n",
    "start = time.time()\n",
    "\n",
    "# Function to run commands and hide output\n",
    "def run_command(command):\n",
    "    subprocess.run(command, shell=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)\n",
    "\n",
    "# Clone RFdiffusion repository directly into RFdiffusionJupyter\n",
    "if not os.path.isdir(\"RFdiffusionJupyter/RFdiffusion\"):\n",
    "    print(\"Cloning RFdiffusion...\")\n",
    "    os.chdir(os.path.expanduser(\"~\"))\n",
    "    os.makedirs(\"RFdiffusionJupyter\", exist_ok=True)\n",
    "    os.chdir(os.path.expanduser(\"RFdiffusionJupyter\"))\n",
    "    run_command(\"git clone https://github.com/sokrypton/RFdiffusion.git\")\n",
    "    os.chdir(\"RFdiffusion\")\n",
    "    run_command(\"wget -qnc https://files.ipd.uw.edu/krypton/ananas\")\n",
    "    run_command(\"chmod +x ananas\")\n",
    "    os.chdir(os.path.expanduser(\"~\"))\n",
    "    print(\"Cloning RFdiffusion and installing dependencies... Done\")\n",
    "\n",
    "# Install necessary packages\n",
    "os.chdir(os.path.expanduser(\"~\"))\n",
    "run_command(\"pip install ipywidgets py3Dmol\")\n",
    "run_command(\"pip install jedi omegaconf hydra-core icecream pyrsistent pynvml decorator\")\n",
    "run_command(\"pip install git+https://github.com/NVIDIA/dllogger#egg=dllogger\")\n",
    "run_command(\"pip install --no-dependencies dgl==2.0.0 -f https://data.dgl.ai/wheels/cu121/repo.html\")\n",
    "run_command(\"pip install --no-dependencies e3nn==0.3.3 opt_einsum_fx\")\n",
    "run_command(\"pip install colabdesign\")\n",
    "run_command(\"pip install -U 'jax[cuda12]'\")\n",
    "os.chdir(os.path.expanduser(\"RFdiffusion\"))\n",
    "os.chdir(\"env/SE3Transformer\")\n",
    "run_command(\"pip install .\")\n",
    "os.chdir(os.path.expanduser(\"~\"))\n",
    "\n",
    "# Download model parameters and unzip if necessary\n",
    "# Check if params are already downloaded\n",
    "if not os.path.isfile(\"RFdiffusionJupyter/RFdiffusion/params/done.txt\"):\n",
    "    print(\"Downloading AF2 params...\")\n",
    "    os.chdir(os.path.expanduser(\"~\"))\n",
    "    params_dir = \"RFdiffusionJupyter/RFdiffusion/params\"\n",
    "    os.makedirs(params_dir, exist_ok=True)\n",
    "    run_command(f\"wget -c https://files.ipd.uw.edu/krypton/schedules.zip -P {params_dir}\")\n",
    "    run_command(f\"wget -c https://storage.googleapis.com/alphafold/alphafold_params_2022-12-06.tar -P {params_dir}\")\n",
    "    run_command(f\"tar -xf {params_dir}/alphafold_params_2022-12-06.tar -C {params_dir}\")\n",
    "    run_command(f\"unzip {params_dir}/schedules.zip -d {params_dir}\")\n",
    "    run_command(f\"touch {params_dir}/done.txt\")\n",
    "\n",
    "\n",
    "if not os.path.isdir(\"RFdiffusionJupyter/RFdiffusion/params/ColabDesign\"):\n",
    "    print(\"Cloning ColabDesign ...\")\n",
    "    os.chdir(os.path.expanduser(\"~/RFdiffusionJupyter/RFdiffusion/params\"))\n",
    "    run_command(\"git clone https://github.com/sokrypton/ColabDesign.git\")\n",
    "    os.chdir(os.path.expanduser(\"~\"))\n",
    "    print(\"Cloning ColabDesign and installing dependencies... Done\")\n",
    "\n",
    "os.chdir(os.path.expanduser(\"~\"))\n",
    "\n",
    "# Check if all models are downloaded into the new /models directory\n",
    "model_files = [\"Base_ckpt.pt\", \"Complex_base_ckpt.pt\", \"Complex_beta_ckpt.pt\"]\n",
    "if not all(os.path.isfile(f\"RFdiffusionJupyter/RFdiffusion/models/{model}\") for model in model_files):\n",
    "    print(\"Downloading model weights...\")\n",
    "    models_dir = \"RFdiffusionJupyter/RFdiffusion/models\"\n",
    "    run_command(f\"wget -c http://files.ipd.uw.edu/pub/RFdiffusion/6f5902ac237024bdd0c176cb93063dc4/Base_ckpt.pt -P {models_dir}\")\n",
    "    run_command(f\"wget -c http://files.ipd.uw.edu/pub/RFdiffusion/e29311f6f1bf1af907f9ef9f44b8328b/Complex_base_ckpt.pt -P {models_dir}\")\n",
    "    run_command(f\"wget -c http://files.ipd.uw.edu/pub/RFdiffusion/f572d396fae9206628714fb2ce00f72e/Complex_beta_ckpt.pt -P {models_dir}\")  \n",
    "    \n",
    "if 'RFdiffusionJupyter/RFdiffusion' not in sys.path:\n",
    "    os.environ[\"DGLBACKEND\"] = \"pytorch\"\n",
    "    sys.path.append('RFdiffusionJupyter/RFdiffusion')\n",
    "\n",
    "os.chdir(os.path.expanduser(\"RFdiffusionJupyter/RFdiffusion\"))\n",
    "\n",
    "# Import modules\n",
    "import json\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import display, HTML\n",
    "import ipywidgets as widgets\n",
    "import py3Dmol\n",
    "\n",
    "# Importing the necessary modules for inference\n",
    "from inference.utils import parse_pdb\n",
    "from colabdesign.rf.utils import get_ca, fix_contigs, fix_partial_contigs, fix_pdb, sym_it\n",
    "from colabdesign.shared.protein import pdb_to_string\n",
    "from colabdesign.shared.plot import plot_pseudo_3D\n",
    "\n",
    "def get_pdb(pdb_code=None):\n",
    "    if pdb_code is None or pdb_code == \"\":\n",
    "        raise ValueError(\"Error: A PDB ID or a path to a PDB file must be provided.\")\n",
    "    if len(pdb_code) == 4:\n",
    "        if not os.path.isfile(f\"{pdb_code}.pdb1\"):\n",
    "            subprocess.run([\"wget\", \"-qnc\", f\"https://files.rcsb.org/download/{pdb_code}.pdb1.gz\"])\n",
    "            subprocess.run([\"gunzip\", f\"{pdb_code}.pdb1.gz\"])\n",
    "        return f\"{pdb_code}.pdb1\"\n",
    "    elif len(pdb_code) == 6 or len(pdb_code) == 10:\n",
    "        subprocess.run([\"wget\", \"-qnc\", f\"https://alphafold.ebi.ac.uk/files/AF-{pdb_code}-F1-model_v3.pdb\"])\n",
    "        return f\"AF-{pdb_code}-F1-model_v3.pdb\"\n",
    "    elif os.path.isfile(pdb_code):\n",
    "        return pdb_code\n",
    "    \n",
    "def run_ananas(pdb_str, path, sym=None):\n",
    "    pdb_filename = f\"RFdiffusionJupyter/RFdiffusion/outputs/{path}/ananas_input.pdb\"\n",
    "    out_filename = f\"RFdiffusionJupyter/RFdiffusion/outputs/{path}/ananas.json\"\n",
    "    with open(pdb_filename, \"w\") as handle:\n",
    "        handle.write(pdb_str)\n",
    "\n",
    "    cmd = f\"./ananas {pdb_filename} -u -j {out_filename}\"\n",
    "    if sym is None:\n",
    "        os.system(cmd)\n",
    "    else:\n",
    "        os.system(f\"{cmd} {sym}\")\n",
    "\n",
    "    # parse results\n",
    "    try:\n",
    "        out = json.loads(open(out_filename, \"r\").read())\n",
    "        results, AU = out[0], out[-1][\"AU\"]\n",
    "        group = AU[\"group\"]\n",
    "        chains = AU[\"chain names\"]\n",
    "        rmsd = results[\"Average_RMSD\"]\n",
    "        print(f\"AnAnaS detected {group} symmetry at RMSD:{rmsd:.3}\")\n",
    "\n",
    "        C = np.array(results['transforms'][0]['CENTER'])\n",
    "        A = [np.array(t[\"AXIS\"]) for t in results['transforms']]\n",
    "\n",
    "        # apply symmetry and filter to the asymmetric unit\n",
    "        new_lines = []\n",
    "        for line in pdb_str.split(\"\\n\"):\n",
    "            if line.startswith(\"ATOM\"):\n",
    "                chain = line[21:22]\n",
    "                if chain in chains:\n",
    "                    x = np.array([float(line[i:(i+8)]) for i in [30, 38, 46]])\n",
    "                    if group[0] == \"c\":\n",
    "                        x = sym_it(x, C, A[0])\n",
    "                    if group[0] == \"d\":\n",
    "                        x = sym_it(x, C, A[1], A[0])\n",
    "                    coord_str = \"\".join([\"{:8.3f}\".format(a) for a in x])\n",
    "                    new_lines.append(line[:30] + coord_str + line[54:])\n",
    "            else:\n",
    "                new_lines.append(line)\n",
    "        return results, \"\\n\".join(new_lines)\n",
    "\n",
    "    except:\n",
    "        return None, pdb_str\n",
    "\n",
    "def run(command, steps, num_designs=1, visual=\"none\"):\n",
    "\n",
    "    def run_command_and_get_pid(command):\n",
    "        pid_file = '/dev/shm/pid'\n",
    "        os.system(f'nohup {command} & echo $! > {pid_file}')\n",
    "        with open(pid_file, 'r') as f:\n",
    "            pid = int(f.read().strip())\n",
    "        os.remove(pid_file)\n",
    "        return pid\n",
    "\n",
    "    def is_process_running(pid):\n",
    "        try:\n",
    "            os.kill(pid, 0)\n",
    "        except OSError:\n",
    "            return False\n",
    "        else:\n",
    "            return True\n",
    "\n",
    "    run_output = widgets.Output()\n",
    "    progress = widgets.FloatProgress(min=0, max=1, description='running', bar_style='info')\n",
    "    display(widgets.VBox([progress, run_output]))\n",
    "\n",
    "    # clear previous run\n",
    "    for n in range(steps):\n",
    "        if os.path.isfile(f\"/dev/shm/{n}.pdb\"):\n",
    "            os.remove(f\"/dev/shm/{n}.pdb\")\n",
    "\n",
    "    pid = run_command_and_get_pid(command)\n",
    "    try:\n",
    "        fail = False\n",
    "        for _ in range(num_designs):\n",
    "\n",
    "            # for each step check if output generated\n",
    "            for n in range(steps):\n",
    "                wait = True\n",
    "                while wait and not fail:\n",
    "                    time.sleep(0.1)\n",
    "                    if os.path.isfile(f\"/dev/shm/{n}.pdb\"):\n",
    "                        pdb_str = open(f\"/dev/shm/{n}.pdb\").read()\n",
    "                        if pdb_str[-3:] == \"TER\":\n",
    "                            wait = False\n",
    "                        elif not is_process_running(pid):\n",
    "                            fail = True\n",
    "                    elif not is_process_running(pid):\n",
    "                        fail = True\n",
    "\n",
    "                if fail:\n",
    "                    progress.bar_style = 'danger'\n",
    "                    progress.description = \"failed\"\n",
    "                    break\n",
    "\n",
    "                else:\n",
    "                    progress.value = (n + 1) / steps\n",
    "                if os.path.exists(f\"/dev/shm/{n}.pdb\"):\n",
    "                    os.remove(f\"/dev/shm/{n}.pdb\")\n",
    "            if fail:\n",
    "                progress.bar_style = 'danger'\n",
    "                progress.description = \"failed\"\n",
    "                break\n",
    "\n",
    "        while is_process_running(pid):\n",
    "            time.sleep(0.1)\n",
    "\n",
    "    except KeyboardInterrupt:\n",
    "        os.kill(pid, signal.SIGTERM)\n",
    "        progress.bar_style = 'danger'\n",
    "        progress.description = \"stopped\"\n",
    "\n",
    "def run_diffusion(contigs, path, pdb=None, iterations=50,\n",
    "                  symmetry=\"none\", order=1, hotspot=None,\n",
    "                  chains=None, add_potential=False, partial_T=\"auto\",\n",
    "                  num_designs=1, use_beta_model=False, visual=\"none\"):\n",
    "\n",
    "    os.chdir(os.path.expanduser(\"~\"))\n",
    "    full_path = f\"RFdiffusionJupyter/RFdiffusion/outputs/{path}\"\n",
    "    os.makedirs(full_path, exist_ok=True)\n",
    "    assert os.path.exists(full_path), \"Output directory was not created!\"\n",
    "    opts = [f\"inference.output_prefix={full_path}\",\n",
    "            f\"inference.num_designs={num_designs}\"]\n",
    "\n",
    "    if chains == \"\":\n",
    "        chains = None\n",
    "\n",
    "    # determine symmetry type\n",
    "    if symmetry in [\"auto\", \"cyclic\", \"dihedral\"]:\n",
    "        if symmetry == \"auto\":\n",
    "            sym, copies = None, 1\n",
    "        else:\n",
    "            sym, copies = {\"cyclic\": (f\"c{order}\", order),\n",
    "                           \"dihedral\": (f\"d{order}\", order * 2)}[symmetry]\n",
    "    else:\n",
    "        symmetry = None\n",
    "        sym, copies = None, 1\n",
    "\n",
    "    # determine mode\n",
    "    contigs = contigs.replace(\",\", \" \").replace(\":\", \" \").split()\n",
    "    is_fixed, is_free = False, False\n",
    "    fixed_chains = []\n",
    "    for contig in contigs:\n",
    "        for x in contig.split(\"/\"):\n",
    "            a = x.split(\"-\")[0]\n",
    "            if a[0].isalpha():\n",
    "                is_fixed = True\n",
    "                if a[0] not in fixed_chains:\n",
    "                    fixed_chains.append(a[0])\n",
    "            if a.isnumeric():\n",
    "                is_free = True\n",
    "    if len(contigs) == 0 or not is_free:\n",
    "        mode = \"partial\"\n",
    "    elif is_fixed:\n",
    "        mode = \"fixed\"\n",
    "    else:\n",
    "        mode = \"free\"\n",
    "\n",
    "    # fix input contigs\n",
    "    if mode in [\"partial\", \"fixed\"]:\n",
    "        pdb_str = pdb_to_string(get_pdb(pdb), chains=chains)\n",
    "        if symmetry == \"auto\":\n",
    "            a, pdb_str = run_ananas(pdb_str, path)\n",
    "            if a is None:\n",
    "                print(f'ERROR: no symmetry detected')\n",
    "                symmetry = None\n",
    "                sym, copies = None, 1\n",
    "            else:\n",
    "                if a[\"group\"][0] == \"c\":\n",
    "                    symmetry = \"cyclic\"\n",
    "                    sym, copies = a[\"group\"], int(a[\"group\"][1:])\n",
    "                elif a[\"group\"][0] == \"d\":\n",
    "                    symmetry = \"dihedral\"\n",
    "                    sym, copies = a[\"group\"], 2 * int(a[\"group\"][1:])\n",
    "                else:\n",
    "                    print(f'ERROR: the detected symmetry ({a[\"group\"]}) not currently supported')\n",
    "                    symmetry = None\n",
    "                    sym, copies = None, 1\n",
    "\n",
    "        elif mode == \"fixed\":\n",
    "            pdb_str = pdb_to_string(pdb_str, chains=fixed_chains)\n",
    "\n",
    "        pdb_filename = f\"{full_path}/input.pdb\"\n",
    "        with open(pdb_filename, \"w\") as handle:\n",
    "            handle.write(pdb_str)\n",
    "\n",
    "        parsed_pdb = parse_pdb(pdb_filename)\n",
    "        opts.append(f\"inference.input_pdb={pdb_filename}\")\n",
    "        if mode in [\"partial\"]:\n",
    "            if partial_T == \"auto\":\n",
    "                iterations = int(80 * (iterations / 200))\n",
    "            else:\n",
    "                iterations = int(partial_T)\n",
    "            opts.append(f\"diffuser.partial_T={iterations}\")\n",
    "            contigs = fix_partial_contigs(contigs, parsed_pdb)\n",
    "        else:\n",
    "            opts.append(f\"diffuser.T={iterations}\")\n",
    "            contigs = fix_contigs(contigs, parsed_pdb)\n",
    "    else:\n",
    "        opts.append(f\"diffuser.T={iterations}\")\n",
    "        parsed_pdb = None\n",
    "        contigs = fix_contigs(contigs, parsed_pdb)\n",
    "\n",
    "    if hotspot is not None and hotspot != \"\":\n",
    "        hotspot = \",\".join(hotspot.replace(\",\", \" \").split())\n",
    "        opts.append(f\"ppi.hotspot_res='[{hotspot}]'\")\n",
    "\n",
    "    # setup symmetry\n",
    "    if sym is not None:\n",
    "        sym_opts = [\"--config-name symmetry\", f\"inference.symmetry={sym}\"]\n",
    "        if add_potential:\n",
    "            sym_opts += [\"'potentials.guiding_potentials=[\\\"type:olig_contacts,weight_intra:1,weight_inter:0.1\\\"]'\",\n",
    "                         \"potentials.olig_intra_all=True\", \"potentials.olig_inter_all=True\",\n",
    "                         \"potentials.guide_scale=2\", \"potentials.guide_decay=quadratic\"]\n",
    "        opts = sym_opts + opts\n",
    "        contigs = sum([contigs] * copies, [])\n",
    "\n",
    "    opts.append(f\"'contigmap.contigs=[{' '.join(contigs)}]'\")\n",
    "    opts += [\"inference.dump_pdb=True\", \"inference.dump_pdb_path='/dev/shm'\"]\n",
    "    if use_beta_model:\n",
    "        opts += [\"inference.ckpt_override_path=RFdiffusionJupyter/Rfdiffusion/models/Complex_beta_ckpt.pt\"]\n",
    "\n",
    "    print(\"mode:\", mode)\n",
    "    print(\"output:\", full_path)\n",
    "    print(\"contigs:\", contigs)\n",
    "\n",
    "    opts_str = \" \".join(opts)\n",
    "    cmd = f\"RFdiffusionJupyter/RFdiffusion/run_inference.py {opts_str}\"\n",
    "    print(cmd)\n",
    "\n",
    "    # RUN\n",
    "    cmd = f\"RFdiffusionJupyter/RFdiffusion/run_inference.py {opts_str}\"\n",
    "    print(f\"Running command: {cmd}\")\n",
    "    run(cmd, iterations, num_designs, visual=visual)\n",
    "\n",
    "    # fix pdbs\n",
    "    for n in range(num_designs):\n",
    "        pdbs = [f\"RFdiffusionJupyter/RFdiffusion/outputs/traj/{path}_{n}_pX0_traj.pdb\",\n",
    "                f\"RFdiffusionJupyter/RFdiffusion/outputs/traj/{path}_{n}_Xt-1_traj.pdb\",\n",
    "                f\"{full_path}_{n}.pdb\"]\n",
    "        for pdb in pdbs:\n",
    "            if not os.path.exists(pdb):\n",
    "                print(f\"WARNING: {pdb} was not generated\")\n",
    "            with open(pdb, \"r\") as handle:\n",
    "                pdb_str = handle.read()\n",
    "            with open(pdb, \"w\") as handle:\n",
    "                handle.write(fix_pdb(pdb_str, contigs))\n",
    "    os.chdir(os.path.expanduser(\"~\"))\n",
    "    return contigs, copies\n",
    "\n",
    "end = time.time()\n",
    "os.chdir(os.path.expanduser(\"~\"))\n",
    "print(f\"Installation took {end-start} seconds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following step is actual diffusion. Parameters you can change are:\n",
    "\n",
    "**name** - name of output files\n",
    "\n",
    "**contigs** - specify what to diffuse\n",
    " - random monomer: contigs = \"x\" with x the length of amino acid chain\n",
    " - binder: specify chain(s) and the relevant amino acids you want to diffuse a binder against as well as the length of the binders amino acid chain; for binding vs a monomer, contigs = \"Asss-eee:x\" with sss as starting amino acid on chain A, eee ending amino acid on chain A and x the length of the binders amino acid chain; for binding vs a multimer contigs=\"Csss-eee:C2sss2-eee2:C3sss3-eee3:x\" for different chains C and starting/ending amino acid respectively\n",
    " - for motif scaffolding or partial diffusion check the original notebook or https://github.com/RosettaCommons/RFdiffusion for instructions\n",
    "\n",
    "**pdb** - PDB ID of the protein you want to design a binder against, if you use the UNIPROT ID you will design vs the AF prediction from https://alphafold-ebi-ac-uk.translate.goog/?_x_tr_sl=en&_x_tr_tl=de&_x_tr_hl=de&_x_tr_pto=sc&_x_tr_hist=true; If you do have a custom PDB file you can also use by specifying the full file path like \"~/RFdiffusion/inputs/someprotein.pdb\"\n",
    "\n",
    "**iterations** [\"25\", \"50\", \"100\", \"150\", \"200\"] - number of iterations for the diffusion of a single backbone, increases runtime but smaller more exact steps towards minimum\n",
    "\n",
    "**hotspot** - preferably hydrophobic amino acids on the specified contig which could increase avidity, as binding is dominated by hydrophobic interactions; specified via hotspot = \"Caaa, ...\" with chain C and the number of the hotspot residue aaa; 3-6 amino acids work best but exact amount should be experimented\n",
    "\n",
    "**num_designs** [\"1\", \"2\", \"4\", \"8\", \"16\", \"32\"] - number of diffused designs, list of possible numbers can be neglected, just specify a different number. \n",
    "\n",
    "**visual**  - removed as there are some problems within Jupyter Lab, keep it at \"none\" otherwise it will break\n",
    "\n",
    "**symmetry** [\"none\", \"auto\", \"cyclic\", \"dihedral\"] - symmetry of diffused protein\n",
    "\n",
    "**order** [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\"] - order of chosen symmetry\n",
    "\n",
    "**chains** - filter input PDB to specified chains\n",
    "\n",
    "**add_potential** [\"True\", \"False\"] - discourage clashes between chains\n",
    "\n",
    "**partial_T** - only used in partial diffusion protocol, check original notebook\n",
    "\n",
    "**use_beta_model** [\"True\", \"False\"] - can be used if you are seeing lots of helices for better SSE balance, however not tested experimentally"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "TuRUfQJZ4vkM"
   },
   "outputs": [],
   "source": [
    "start = time.time()\n",
    "import torch\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "name = \"SpyTagScaffold\" \n",
    "contigs = \"B:100\" \n",
    "pdb = \"4MLI\" \n",
    "iterations = 50 \n",
    "hotspot = \"\" \n",
    "num_designs = 30\n",
    "visual = \"none\" \n",
    "symmetry = \"cyclic\"\n",
    "order = 3 \n",
    "chains = \"\" \n",
    "add_potential = True \n",
    "partial_T = \"auto\" \n",
    "use_beta_model = False \n",
    "# determine where to save\n",
    "path = name\n",
    "os.chdir(os.path.expanduser(\"~\"))\n",
    "while os.path.exists(f\"RFdiffusionJupyter/RFdiffusion/outputs/{path}_0.pdb\"):\n",
    "  path = name + \"_\" + ''.join(random.choices(string.ascii_lowercase + string.digits, k=5))\n",
    "\n",
    "flags = {\"contigs\":contigs,\n",
    "         \"pdb\":pdb,\n",
    "         \"order\":order,\n",
    "         \"iterations\":iterations,\n",
    "         \"symmetry\":symmetry,\n",
    "         \"hotspot\":hotspot,\n",
    "         \"path\":path,\n",
    "         \"chains\":chains,\n",
    "         \"add_potential\":add_potential,\n",
    "         \"num_designs\":num_designs,\n",
    "         \"use_beta_model\":use_beta_model,\n",
    "         \"visual\":visual,\n",
    "         \"partial_T\":partial_T}\n",
    "\n",
    "for k,v in flags.items():\n",
    "  if isinstance(v,str):\n",
    "    flags[k] = v.replace(\"'\",\"\").replace('\"','')\n",
    "    \n",
    "os.chdir(os.path.expanduser(\"~\"))\n",
    "contigs, copies = run_diffusion(**flags)\n",
    "end = time.time()\n",
    "print(f\"Diffusion took {end-start} seconds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " The following code allows you to display your diffused backbones with the specified contig in three dimensions. Here you can specify three parameters:\n",
    " \n",
    " **animate** - [\"none\", \"movie\", \"interactive\"]\n",
    " \n",
    " **color** - [\"rainbow\", \"chain\", \"plddt\"]\n",
    " \n",
    " **dpi** - [\"100\", \"200\", \"400\"]\n",
    " \n",
    " **design** - specify the design you want to model, starting from 0 which is also the default and will be displayed if you do not specify anthing!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "wqEi03_qi_g2"
   },
   "outputs": [],
   "source": [
    "import torch\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "animate = \"none\" \n",
    "color = \"chain\"\n",
    "denoise = True\n",
    "dpi = 100\n",
    "design = 2\n",
    "from colabdesign.shared.plot import pymol_color_list\n",
    "from colabdesign.rf.utils import get_ca, get_Ls, make_animation\n",
    "from string import ascii_uppercase,ascii_lowercase\n",
    "alphabet_list = list(ascii_uppercase+ascii_lowercase)\n",
    "\n",
    "def plot_pdb(num=0):\n",
    "  os.chdir(os.path.expanduser(\"~\"))  \n",
    "  if denoise:\n",
    "    pdb_traj = f\"RFdiffusionJupyter/RFdiffusion/outputs/traj/{path}_{num}_pX0_traj.pdb\"\n",
    "  else:\n",
    "    pdb_traj = f\"RFdiffusionJupyter/RFdiffusion/outputs/traj/{path}_{num}_Xt-1_traj.pdb\"\n",
    "  if animate in [\"none\",\"interactive\"]:\n",
    "    hbondCutoff = 4.0\n",
    "    view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js')\n",
    "    if animate == \"interactive\":\n",
    "      pdb_str = open(pdb_traj,'r').read()\n",
    "      view.addModelsAsFrames(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
    "    else:\n",
    "      pdb = f\"RFdiffusionJupyter/RFdiffusion/outputs/{path}_{num}.pdb\"\n",
    "      pdb_str = open(pdb,'r').read()\n",
    "      view.addModel(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
    "    if color == \"rainbow\":\n",
    "      view.setStyle({'cartoon': {'color':'spectrum'}})\n",
    "    elif color == \"chain\":\n",
    "      for n,chain,c in zip(range(len(contigs)),\n",
    "                              alphabet_list,\n",
    "                              pymol_color_list):\n",
    "          view.setStyle({'chain':chain},{'cartoon': {'color':c}})\n",
    "    else:\n",
    "      view.setStyle({'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':0.5,'max':0.9}}})\n",
    "    view.zoomTo()\n",
    "    if animate == \"interactive\":\n",
    "      view.animate({'loop': 'backAndForth'})\n",
    "    view.show()\n",
    "  else:\n",
    "    Ls = get_Ls(contigs)\n",
    "    xyz, bfact = get_ca(pdb_traj, get_bfact=True)\n",
    "    xyz = xyz.reshape((-1,sum(Ls),3))[::-1]\n",
    "    bfact = bfact.reshape((-1,sum(Ls)))[::-1]\n",
    "    if color == \"chain\":\n",
    "      display(HTML(make_animation(xyz, Ls=Ls, dpi=dpi, ref=-1)))\n",
    "    elif color == \"rainbow\":\n",
    "      display(HTML(make_animation(xyz, dpi=dpi, ref=-1)))\n",
    "    else:\n",
    "      display(HTML(make_animation(xyz, plddt=bfact*100, dpi=dpi, ref=-1)))\n",
    "\n",
    "\n",
    "if num_designs > 1:\n",
    "    plot_pdb(design)\n",
    "else:\n",
    "  plot_pdb()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The final step is generating a sequence for the diffused backbone with **ProteinMPNN**, model it with **AF2** and evaluate the designs depending on interaction predicted Aligned Error and complex root mean square deviation (i_pAE and rmsd --> best case smaller than ten)! Here you have a number of parameters which you can change: \n",
    "\n",
    "***ProteinMPNN settings:***\n",
    "\n",
    "**num_seqs** [\"1\", \"2\", \"4\", \"8\", \"16\", \"32\", \"64\"] - number of designed sequences per diffused backbone\n",
    "\n",
    "**mpnn_sampling_temp** [\"0.0001\", \"0.1\", \"0.15\", \"0.2\", \"0.25\", \"0.3\", \"0.5\", \"1.0\"] - states the diversity of sampled sequences with higher numbers resulting in more diverse sequences\n",
    "\n",
    "**rm_aa** ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y'] - allows you to remove certain amino acids\n",
    "\n",
    "**use_solubleMPNN** [\"True\", \"False\"] - there sadly where some issues in Jupyter Lab when trying to include this, so you have to remove the arguments (cursive) during initialisation of the mpnn model in \n",
    "RFdiffusionJupyter/RFdiffusion/params/ColabDesign/colabdesign/rf/designability_test.py\n",
    "print(\"running proteinMPNN...\")\n",
    "  sampling_temp = 0.1\n",
    "  mpnn_model = mk_mpnn_model(*weights=\"soluble\" if o.use_soluble else \"original\"*)\n",
    "When you want to use the soluble weights, you must then cd into: RFdiffusionJupyter/RFdiffusion/params/ColabDesign/colabdesign/mpnn/model.py and change the standard for the 'weights' variable to either 'original' or 'soluble'within the __init__ method of the class.\n",
    "\n",
    "class mk_mpnn_model():\n",
    "  def __init__(self, model_name=\"v_48_020\",\n",
    "               backbone_noise=0.0, dropout=0.0,\n",
    "               seed=None, verbose=False, weights=\"original\"): # weights can be set to either original or soluble\n",
    "'\n",
    "\n",
    "***AF2 settings:***\n",
    "\n",
    "**initial_guess** [\"True\", \"False\"] and **num_recycles** [\"0\", \"1\", \"2\", \"3\", \"6\", \"12\"] - for **binder** design, it is recommended to choose `initial_guess=True num_recycles=3´\n",
    "\n",
    "**use_multimer** [\"True\", \"False\"] - use AlphaFold Multimer v3 params for prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "rES3p-q6j4tc"
   },
   "outputs": [],
   "source": [
    "start = time.time()\n",
    "import torch\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "num_seqs = 8 \n",
    "num_recycles = 3\n",
    "mpnn_sampling_temp = 0.1 \n",
    "rm_aa = \"C\" \n",
    "use_solubleMPNN = False\n",
    "initial_guess = True\n",
    "use_multimer = False \n",
    "os.chdir(os.path.expanduser(\"~\"))\n",
    "if not os.path.isfile(\"RFdiffusionJupyter/RFdiffusion/params/done.txt\"):\n",
    "  print(\"downloading AlphaFold params...\")\n",
    "  while not os.path.isfile(\"params/done.txt\"):\n",
    "    time.sleep(5)\n",
    "\n",
    "contigs_str = \":\".join(contigs)\n",
    "opts = [f\"--pdb=../outputs/{path}_0.pdb\",\n",
    "        f\"--loc=../outputs/{path}\",\n",
    "        f\"--contig={contigs_str}\",\n",
    "        f\"--copies={copies}\",\n",
    "        f\"--num_seqs={num_seqs}\",\n",
    "        f\"--num_recycles={num_recycles}\",\n",
    "        f\"--rm_aa={rm_aa}\",\n",
    "        f\"--mpnn_sampling_temp={mpnn_sampling_temp}\",\n",
    "        f\"--num_designs={num_designs}\"]\n",
    "if initial_guess: opts.append(\"--initial_guess\")\n",
    "if use_multimer: opts.append(\"--use_multimer\")\n",
    "if use_solubleMPNN: opts.append(\"--use_soluble\")\n",
    "opts = ' '.join(opts)\n",
    "os.chdir(os.path.expanduser(\"~/RFdiffusionJupyter/RFdiffusion/params\"))\n",
    "!python ColabDesign/colabdesign/rf/designability_test.py {opts}\n",
    "end = time.time()\n",
    "print(f\"Sequence design and ranking took {end - start} seconds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With this final cell you can display the best sequence for each design. If you want to specify of which design you want to see the best sequence, change the design variable to to according design number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "DUNKRBNSvk6_"
   },
   "outputs": [],
   "source": [
    "import torch\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "import py3Dmol\n",
    "os.chdir(os.path.expanduser(\"~\"))\n",
    "design = \"best\"\n",
    "def plot_pdb(num = \"best\"):\n",
    "  if num == \"best\":\n",
    "    with open(f\"RFdiffusionJupyter/RFdiffusion/outputs/{path}/best.pdb\",\"r\") as f:\n",
    "      # REMARK 001 design {m} N {n} RMSD {rmsd}\n",
    "      info = f.readline().strip('\\n').split()\n",
    "    num = info[3]\n",
    "  hbondCutoff = 4.0\n",
    "  view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js')\n",
    "  pdb_str = open(f\"RFdiffusionJupyter/RFdiffusion/outputs/{path}_{num}.pdb\",'r').read()\n",
    "  view.addModel(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
    "  pdb_str = open(f\"RFdiffusionJupyter/RFdiffusion/outputs/{path}/best_design{num}.pdb\",'r').read()\n",
    "  view.addModel(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
    "\n",
    "  view.setStyle({\"model\":0},{'cartoon':{}}) #: {'colorscheme': {'prop':'b','gradient': 'roygb','min':0,'max':100}}})\n",
    "  view.setStyle({\"model\":1},{'cartoon':{'colorscheme': {'prop':'b','gradient': 'roygb','min':0,'max':100}}})\n",
    "  view.zoomTo()\n",
    "  view.show()\n",
    "\n",
    "if num_designs>1:\n",
    "    plot_pdb(design)\n",
    "else:\n",
    "    plot_pdb(num)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
